File: tests/client-gen/example_program_gen/types/bar_struct.py
import typing
from construct import Container
import borsh_construct as borsh


class BarStructFields(typing.TypedDict):
    some_field: bool
    other_field: int


class BarStructJSON(typing.TypedDict):
    some_field: bool
    other_field: int


class BarStruct(object):
    layout = borsh.CStruct("some_field" / borsh.Bool, "other_field" / borsh.U8)

    def __init__(self, fields: BarStructFields) -> None:
        self.some_field = fields["some_field"]
        self.other_field = fields["other_field"]

    @classmethod
    def from_decoded(cls, obj: Container) -> "BarStruct":
        return cls(
            BarStructFields(some_field=obj.some_field, other_field=obj.other_field)
        )

    @classmethod
    def to_encodable(cls, fields: BarStructFields) -> dict[str, typing.Any]:
        return {
            "some_field": fields["some_field"],
            "other_field": fields["other_field"],
        }

    def to_json(self) -> BarStructJSON:
        return {"some_field": self.some_field, "other_field": self.other_field}

    @classmethod
    def from_json(cls, obj: BarStructJSON) -> "BarStruct":
        return cls(
            BarStructJSON(some_field=obj["some_field"], other_field=obj["other_field"])
        )

File: ts-reference/tests/example-program-gen/exp/types/BarStruct.ts
import { PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh"

export interface BarStructFields {
  someField: boolean
  otherField: number
}

export interface BarStructJSON {
  someField: boolean
  otherField: number
}

export class BarStruct {
  readonly someField: boolean
  readonly otherField: number

  constructor(fields: BarStructFields) {
    this.someField = fields.someField
    this.otherField = fields.otherField
  }

  static layout(property?: string) {
    return borsh.struct(
      [borsh.bool("someField"), borsh.u8("otherField")],
      property
    )
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  static fromDecoded(obj: any) {
    return new BarStruct({
      someField: obj.someField,
      otherField: obj.otherField,
    })
  }

  static toEncodable(fields: BarStructFields) {
    return {
      someField: fields.someField,
      otherField: fields.otherField,
    }
  }

  toJSON(): BarStructJSON {
    return {
      someField: this.someField,
      otherField: this.otherField,
    }
  }

  static fromJSON(obj: BarStructJSON): BarStruct {
    return new BarStruct({
      someField: obj.someField,
      otherField: obj.otherField,
    })
  }

  toEncodable() {
    return BarStruct.toEncodable(this)
  }
}
